### PR TITLE
NDRS-575: Add Subcommand to retrieve documentation for RPCs

### PR DESCRIPTION
**Pull request by darthsiroftardis**
REF: https://casperlabs.atlassian.net/browse/NDRS-575

I have added a subcommand called `docs_get_rpcs` that displays all currently supported RPC, with documentation on each RPC. I have attached a text file to this PR with the sample output to be reviewed. I asked both Mateusz and Henry to join in the review as there maybe some inconsistencies in the representation of certain RPCs that I may not be aware of. 

*Sample Output*
[ndrs_575_output.txt](https://github.com/CasperLabs/casper-node/files/5537920/ndrs_575_output.txt)

### CODE REVIEW

**Comment by goral09** [comment]
I haven't used the RPC at all, nor have I written the code. I think it would take too much time to provide and valuable input. I would like to unassign myself from the review.

**Review by Fraser999 — CHANGES_REQUESTED** [review]
Please try and not have `cargo fmt` runs as separate commits.  You should be able to set up your IDE or have a post-commit hook to avoid this.  Also, you should always run `make check-rs` before creating a PR or committing to a PR to try and cut down on the number of commits and failed CI runs.

Normally `lazy_static` instances are defined at the top of a source file, in the same way that `const` values are.  I'd be fine with these living inside their respective `doc_example` methods though, e.g.

```rust
impl DocExample for NodeId {
    fn doc_example() -> &'static Self {
        lazy_static! {
            static ref NODE_ID: NodeId =
                NodeId::Tls(KeyFingerprint::from([1u8; KeyFingerprint::LENGTH]));
        }
        &*NODE_ID
    }
}
```

There are two major blockers to merging this PR:
1. The docs RPC response is missing details for the `state_get_item` RPC
2. The response needs to be constant across sessions and nodes.  Please at least check this is true for sessions before asking for re-review.  Run a vault, output the response via a client to a file, restart the vault, output to a different file, and finally check the files are identical (modulo the RPC ID unless you specify one via the command line).

File: node/src/components/chainspec_loader.rs [hunk 1]
```diff
@@ -85,6 +87,19 @@ impl ChainspecInfo {
     }
 }
 
+lazy_static! {
+    static ref CHAINSPECINFO: ChainspecInfo = ChainspecInfo {
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/chainspec_loader.rs:91]: Renamed

File: node/src/components/rpc_server/rpcs/docs.rs [hunk 2]
```diff
@@ -0,0 +1,395 @@
+//! RPCs related to finding information about currently supported RPCs.
+
+use std::{
+    collections::HashMap,
+    net::{IpAddr, Ipv4Addr, SocketAddr},
+};
+
+use casper_types::U512;
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/docs.rs:8]: Moved

File: node/src/components/rpc_server/rpcs/docs.rs [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+    types::{
+        json_compatibility::{AuctionState, ExecutionResult},
+        Block, BlockHash, Deploy, GetStatusResult, NodeId, PeersMap, StatusFeed,
+    },
+};
+
+lazy_static! {
+    static ref MERKLE_PROOF: String = String::from("01000000006ef2e0949ac76e55812421f755abe129b6244fe7168b77f47a72536147614625016ef2e0949ac76e55812421f755abe129b6244fe7168b77f47a72536147614625000000003529cde5c621f857f75f3810611eb4af3f998caaa9d4a3413cf799f99c67db0307010000006ef2e0949ac76e55812421f755abe129b6244fe7168b77f47a7253614761462501010102000000006e06000000000074769d28aac597a36a03a932d4b43e4f10bf0403ee5c41dd035102553f5773631200b9e173e8f05361b681513c14e25e3138639eb03232581db7557c9e8dbbc83ce94500226a9a7fe4f2b7b88d5103a4fc7400f02bf89c860c9ccdd56951a2afe9be0e0267006d820fb5676eb2960e15722f7725f3f8f41030078f8b2e44bf0dc03f71b176d6e800dc5ae9805068c5be6da1a90b2528ee85db0609cc0fb4bd60bbd559f497a98b67f500e1e3e846592f4918234647fca39830b7e1e6ad6f5b7a99b39af823d82ba1873d000003000000010186ff500f287e9b53f823ae1582b1fa429dfede28015125fd233a31ca04d5012002015cc42669a55467a1fdf49750772bfc1aed59b9b085558eb81510e9b015a7c83b0301e3cf4a34b1db6bfa58808b686cb8fe21ebe0c1bcbcee522649d2b135fe510fe3");
+
+
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/docs.rs:46]: Changed it one blank line

File: node/src/components/rpc_server/rpcs/docs.rs [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+    static ref DOCS_RPC_RESULT: GetRpcsResult = {
+        let mut result = GetRpcsResult {
+            api_version: CLIENT_API_VERSION.clone(),
+            rpcs:  vec![],
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/docs.rs:50]: Removed extra whitespace

File: node/src/components/rpc_server/rpcs/docs.rs [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+        };
+
+        // Setup PutDeploy example.
+           let deploy = Deploy::doc_example();
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/docs.rs:54]: Formatted separately, and re-inserted back into `lazy_static`

File: node/src/components/rpc_server/rpcs/docs.rs [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+           let response_result = PutDeployResult {
+                api_version: CLIENT_API_VERSION.clone(),
+                deploy_hash: *deploy.id()
+            };
+            let request_params = PutDeployParams {
+                deploy: deploy.clone()
+            };
+            result.push_with_params::<PutDeploy>(
+                "Creates a Deploy and sends it to the network for execution.",
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/docs.rs:63]: I am reusing the documentation in `client/lib.rs`, should I change it anyway

File: client/lib/lib.rs [hunk 7]
```diff
@@ -309,6 +309,11 @@ pub fn get_auction_info(maybe_rpc_id: &str, node_address: &str, verbose: bool) -
     RpcCall::new(maybe_rpc_id, node_address, verbose)?.get_auction_info()
 }
 
+/// Retrieves all currently supported RPCS.
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply client/lib/lib.rs:312]: Added additional documentation within the client

File: client/src/docs.rs [hunk 8]
```diff
@@ -0,0 +1 @@
+mod get;
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply client/src/docs.rs:1]: Treated similar to `get_auction_info`

File: node/src/components/rpc_server/rpcs/docs.rs [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+                None,
+                "https://docs.rs/casper-node/latest/casper_node/rpcs/state/struct.GetAuctionInfoResult.html",
+                None,
+                response_result,
+            );
+
+        result
+    };
+}
+/// A trait to generate static hardcode representations of data structures to present for RPC calls.
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/docs.rs:229]: Changed documentation

File: node/src/components/rpc_server/rpcs/docs.rs [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+pub trait DocExample {
+    /// Generate a hardcoded, possibly invalid example of the requested data structure.
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/docs.rs:231]: Changed documentation

File: node/src/components/rpc_server/rpcs/docs.rs [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+pub struct GetRpcsResult {
+    /// The RPC API Version.
+    api_version: Version,
+    /// The List of supported RPCs.
+    rpcs: Vec<RpcDocs>,
+}
+
+impl GetRpcsResult {
+    #[allow(clippy::too_many_arguments)]
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/docs.rs:276]: Removed

File: node/src/components/rpc_server/rpcs/docs.rs [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+        notes: Option<&str>,
+        request_params_docs_url: &str,
+        request_params_notes: Option<&str>,
+        request_params_example: T::RequestParams,
+        response_result_docs_url: &str,
+        response_result_notes: Option<&str>,
+        response_result_example: T::ResponseResult,
+    ) {
+        let request_params = RequestParams {
+            requirement: String::from("params must be presenet"),
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/docs.rs:289]: Corrected spelling

File: node/src/components/rpc_server/rpcs/docs.rs [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+    example: Value,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+/// The struct containing the documentation for the RPCs.
+pub struct RpcDocs {
+    method: String,
+    summary: String,
+    notes: Option<String>,
+    request_params: Option<RequestParams>,
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/components/rpc_server/rpcs/docs.rs:262]: Removed `Option` and changed to simply, `RequestParams`

File: node/src/crypto/asymmetric_key.rs [hunk 14]
```diff
@@ -375,6 +377,16 @@ impl Debug for SecretKey {
     }
 }
 
+lazy_static! {
+    static ref ED25519_KEY: SecretKey = SecretKey::generate_ed25519();
+}
+
+impl DocExample for SecretKey {
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/crypto/asymmetric_key.rs:384]: Changed it a `doc_example` that results a consistent value across multiple node sessions

File: node/src/crypto/hash.rs [hunk 15]
```diff
@@ -13,7 +13,7 @@ use blake2::{
 use datasize::DataSize;
 use hex_buffer_serde::{Hex, HexForm};
 use hex_fmt::HexFmt;
-#[cfg(test)]
+
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/crypto/hash.rs:16]: Yes, changed accordingly

File: node/src/types/block.rs [hunk 16]
```diff
@@ -351,6 +374,34 @@ impl FinalizedBlock {
     }
 }
 
+lazy_static! {
+    static ref FINALIZEDBLOCK: FinalizedBlock = {
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/types/block.rs:378]: Renamed

File: node/src/types/block.rs [hunk 17]
```diff
@@ -237,6 +236,30 @@ impl FromBytes for EraEnd {
     }
 }
 
+lazy_static! {
+    static ref ERAEND: EraEnd = {
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/types/block.rs:240]: Renamed

File: node/src/types/block.rs [hunk 18]
```diff
@@ -351,6 +374,34 @@ impl FinalizedBlock {
     }
 }
 
+lazy_static! {
+    static ref FINALIZEDBLOCK: FinalizedBlock = {
+        let deploy_hashes = vec![DeployHash::new(Digest::doc_example())];
+        let random_bit = true;
+        let proto_block = ProtoBlock::new(deploy_hashes, random_bit);
+        let timestamp = Timestamp::now();
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/types/block.rs:382]: Constant values are being generated across multiple node sessions

File: node/src/types/deploy.rs [hunk 19]
```diff
@@ -447,6 +449,62 @@ impl Deploy {
     }
 }
 
+lazy_static! {
+    static ref STORECONTRACTNAME: ExecutableDeployItem = {
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/types/deploy.rs:453]: Removed `lazy_static` for `STORED_CONTRACT_BY_NAME`

File: node/src/types/deploy.rs [hunk 20]
```diff
@@ -447,6 +449,62 @@ impl Deploy {
...
+        let entry_point = String::from("example_entry_point");
+        let args = Digest::doc_example().to_vec();
+
+        ExecutableDeployItem::StoredContractByName {
+            name,
+            entry_point,
+            args,
+        }
+    };
+    static ref MODULEBYTES: ExecutableDeployItem = {
```
[the remark this answers is not in the document]
Comment (darthsiroftardis) [reply node/src/types/deploy.rs:464]: Removed `lazy_static`

**Comment by darthsiroftardis** [comment]
PR Closed, refer to a PR for [NDRS-591](https://github.com/CasperLabs/casper-node/pull/533/commits/8cf853b14d6610955be8b1519240a56af16d9ff5)

File: node/src/components/rpc_server/rpcs/docs.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Fraser999) [reply node/src/components/rpc_server/rpcs/docs.rs]: Don't worry about changing this now.  We'll give these all an overhaul once we decide upon a consistent style.  We'll use the modified "summary" strings from the handwritten doc we created.

**Pull request closed without merging**